Fix app status report export issue

AmitPhulera opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
https://dimagi.atlassian.net/browse/SAAS-16642

Let us consider a web user has access to multiple domains and they are working with/updating apps/making submissions in both the domains.
Their reporting metadata will have information from all the domains. 
Now if we try to get App Status report for one of the domain that this user is part of but their last submission belongs to other domain. The code right now will try to fetch app name for the app in other domain which will fail with a 404. The failure is demonstrated in the [github action run](https://github.com/dimagi/commcare-hq/actions/runs/14332115375/job/40170323307?pr=35882#step:5:9386)

The fix here skips those entries from app status reports where the last activity for user is on other domain.

Apart from that the PR adds tests for AppStatusReport. I have not added test for the specific scenario but just tried to cover the functionality of the report.

## Safety Assurance
Very small change, fixes 404 coming in app status reports.
### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests



<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev commented on corehq/apps/app_manager/exceptions.py:
```diff
@@ -204,3 +204,10 @@ class MobileUCRTooLargeException(CannotRestoreException):
+
+class AppInDifferentDomainException(AppManagerException):
+    """
+    This exception is raised when an app is requested from a domain that is different from the domain of the app.
```
This docstring references two different apps. Which is the second app referring to? Seems like there is some missing context.

jingcheng16 replied:
How about update the docstring to be "This exception is raised when an app is not in the targeted domain" ?

millerdev commented on corehq/apps/app_manager/dbaccessors.py:
https://github.com/dimagi/commcare-hq/pull/35882/commits/862ab5a4d0b711fbe19f62a59a1d2aa9f6df209d could have been combined with https://github.com/dimagi/commcare-hq/pull/35882/commits/cb4a8b333cd9dbb9039c5d4fc8a57d82185f2409. Was there a reason you kept them separate?

millerdev commented on corehq/apps/reports/standard/deployments.py:
```diff
@@ -377,7 +378,13 @@ def process_rows(self, users, fmt_for_export=False):
                 if last_build.get('app_id'):
-                    app_name = self.get_app_name(last_build['app_id'])
+                    try:
+                        # For web users, who are hopping into multiple apps, in different domains,
+                        # we should not fail the whole report if one of the app names is not found
+                        # We should skip the entry for that user as it did not belong to the requested domain
```
```suggestion
                        # For web users who are hopping into multiple apps in different domains,
                        # we should not fail the whole report if one of the app names is not found.
                        # We should skip the entry for that user as it did not belong to the requested domain.
```

millerdev commented on corehq/apps/app_manager/views/forms.py:
```diff
@@ -977,7 +978,10 @@ def get_form_datums(request, domain, app_id):
+    try:
+        app = get_app(domain, app_id)
+    except AppInDifferentDomainException as e:
+        raise Http404(str(e))
```
Should `str(e)` be translated?

millerdev commented on corehq/apps/reports/tests/test_application_status_report.py:
```diff
@@ -0,0 +1,465 @@
+        return self._id
+
+
+class TestApplicationStatusReportData(TestCase):
```
Why does this subclass `TestCase`? It appears that this class only exists for the `get_user_data` function which is called by tests in other classes. Consider moving that function out of this class and to the bottom of the module after tests since it is a helper method.

millerdev commented on corehq/apps/reports/tests/test_application_status_report.py:
```diff
@@ -0,0 +1,465 @@
+
+    @staticmethod
+    def get_user_data(domain):
+        def format_datetime(dt):
+            return json_format_datetime(dt)
```
Why is this function needed? `json_format_datetime` works fine on its own.

millerdev reviewed (changes requested):
How thoroughly did you review the [clearly AI-generated](https://github.com/dimagi/commcare-hq/blob/820073eca0b24a19f67e8331e4b29106c73f126b/corehq/apps/reports/tests/test_application_status_report.py#L36-L37) test code? I'd prefer not to spend more time reviewing until you've done a cleanup pass to ensure that the tests actually make sense, and please remove all superfluous comments.

jingcheng16 commented on corehq/apps/reports/tests/test_application_status_report.py:
```diff
@@ -0,0 +1,465 @@
+        # Now create users with proper metadata
+        for user_data in cls.user_data:
+            # Create the user properly
+            user = CommCareUser(user_data)
```
Since one user have access to apps in different domain can only happen for WebUser, does it make more sense to make it WebUser in the test?

jingcheng16 commented:
Overall looks good to me. Just a nit, I was expecting to see a test, with a web user whose last submission is for another domain. This is to test what was fixed in this PR.

AmitPhulera committed:
add test to demonstrate issue where user object with app from different domain will mess the app status report

AmitPhulera committed:
add fix for the issue describe in previous commit

AmitPhulera committed:
keep same behaviour for exiting functionality

AmitPhulera commented:
Hey @millerdev and @jingcheng16 !

I have added the test case for the issue that this PR was fixing and removed all superfluous tests. It required a lot of setup so the test has a lot of boilerplate. Couldn't figure out a better way to go about it 😅 

I have redone this PR and addressed the feedback that you guys have shared on it earlier.
